File: rust/herb-config/src/config_schema.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use crate::severity::{Severity, SeverityConfig};

#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct FilesConfig {
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub include: Option<Vec<String>>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub exclude: Option<Vec<String>>,
}

#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct RuleConfig {
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub enabled: Option<bool>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub severity: Option<SeverityConfig>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub auto_correct: Option<bool>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub include: Option<Vec<String>>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub only: Option<Vec<String>>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub exclude: Option<Vec<String>>,
}

#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct LinterConfig {
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub enabled: Option<bool>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub fail_level: Option<Severity>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub log_level: Option<Severity>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub include: Option<Vec<String>>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub exclude: Option<Vec<String>>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub rules: Option<HashMap<String, RuleConfig>>,
}

#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct RewriterConfig {
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub pre: Option<Vec<String>>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub post: Option<Vec<String>>,
}

#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct FormatterConfig {
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub enabled: Option<bool>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub include: Option<Vec<String>>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub exclude: Option<Vec<String>>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub indent_width: Option<usize>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub max_line_length: Option<usize>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub rewriter: Option<RewriterConfig>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Framework {
  Ruby,
  ActionView,
  Hanami,
  Sinatra,
}

impl Framework {
  pub const ALL: [Framework; 4] = [Framework::Ruby, Framework::ActionView, Framework::Hanami, Framework::Sinatra];

  pub fn as_str(&self) -> &'static str {
    match self {
      Framework::Ruby => "ruby",
      Framework::ActionView => "actionview",
      Framework::Hanami => "hanami",
      Framework::Sinatra => "sinatra",
    }
  }

  pub fn label(&self) -> &'static str {
    match self {
      Framework::Ruby => "Ruby",
      Framework::ActionView => "Action View",
      Framework::Hanami => "Hanami",
      Framework::Sinatra => "Sinatra",
    }
  }

  pub fn description(&self) -> &'static str {
    match self {
      Framework::Ruby => "plain ERB templates, with no framework helpers in scope",
      Framework::ActionView => "Action View templates, with their helpers, partials, and strict locals",
      Framework::Hanami => "Hanami views, with their parts and helpers",
      Framework::Sinatra => "Sinatra templates, with their helpers",
    }
  }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum TemplateEngine {
  Erubi,
  Erb,
  Herb,
}

pub type EngineConfig = serde_yaml::Mapping;

#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct HerbConfigOptions {
  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub files: Option<FilesConfig>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub engine: Option<EngineConfig>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub linter: Option<LinterConfig>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub formatter: Option<FormatterConfig>,
}

#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct HerbConfig {
  #[serde(default)]
  pub version: String,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub framework: Option<Framework>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub template_engine: Option<TemplateEngine>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub files: Option<FilesConfig>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub engine: Option<EngineConfig>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub linter: Option<LinterConfig>,

  #[serde(default, skip_serializing_if = "Option::is_none")]
  pub formatter: Option<FormatterConfig>,
}

File: javascript/packages/config/src/config-schema.ts
import { z } from "zod"

import { DIAGNOSTIC_SEVERITIES } from "@herb-tools/core"

import type { DiagnosticSeverity } from "@herb-tools/core"

export const SeveritySchema = z.enum(DIAGNOSTIC_SEVERITIES)

export const SeverityConfigSchema = z.union([
  SeveritySchema,
  z.object({ editor: SeveritySchema, cli: SeveritySchema }).strict()
])

export const FilesConfigSchema = z.object({
  include: z.array(z.string()).optional().describe("Additional glob patterns to include beyond defaults (e.g., ['**/*.xml.erb', 'custom/**/*.html'])"),
  exclude: z.array(z.string()).optional().describe("Glob patterns to exclude (e.g., ['node_modules/**/*', 'vendor/**/*', '**/*.html.erb'])"),
}).strict().optional()

export const FRAMEWORKS = {
  ruby: "Ruby",
  actionview: "Action View",
  hanami: "Hanami",
  sinatra: "Sinatra",
} as const

export const FRAMEWORK_NAMES = Object.keys(FRAMEWORKS) as (keyof typeof FRAMEWORKS)[]

export const ENVIRONMENT_NAMES = ["cli", "browser"] as const

const RuleConfigBaseSchema = z.object({
  enabled: z.boolean().optional().describe("Whether the rule is enabled"),
  severity: SeverityConfigSchema.optional().describe("Severity level for the rule"),
  frameworks: z.array(z.enum(FRAMEWORK_NAMES)).optional().describe("Frameworks this rule applies to (defaults to every framework)"),
  include: z.array(z.string()).optional().describe("Additional glob patterns to include for this rule (additive, ignored when 'only' is present)"),
  only: z.array(z.string()).optional().describe("Only apply this rule to files matching these glob patterns (overrides all 'include' patterns)"),
  exclude: z.array(z.string()).optional().describe("Don't apply this rule to files matching these glob patterns"),
  environments: z.array(z.enum(ENVIRONMENT_NAMES)).optional().describe("Where this rule runs: 'cli' for templates read from source, 'browser' for a rendered page read from a live DOM. Defaults to ['cli'], so a rule only runs against a rendered page when it says it can"),
})

export const RuleConfigSchema = RuleConfigBaseSchema.optional()

export const LinterConfigSchema = z.object({
  enabled: z.boolean().optional().describe("Whether the linter is enabled"),
  failLevel: SeveritySchema.optional().describe("Exit with error code when diagnostics of this severity or higher are present (e.g., 'warning' will fail on warnings and errors)"),
  logLevel: SeveritySchema.optional().describe("Only report diagnostics of this severity or higher (e.g., 'warning' hides info and hint diagnostics from the output and from CI annotations)"),
  include: z.array(z.string()).optional().describe("Additional glob patterns to include beyond defaults (e.g., ['**/*.xml.erb', 'custom/**/*.html'])"),
  exclude: z.array(z.string()).optional().describe("Glob patterns to exclude from linting"),
  rules: z.record(z.string(), RuleConfigBaseSchema).optional().describe("Per-rule configuration"),
}).strict().optional()

const RewriterConfigSchema = z.object({
  pre: z.array(z.string()).optional().describe("Pre-format rewriters to run (in order) before formatting the AST"),
  post: z.array(z.string()).optional().describe("Post-format rewriters to run (in order) after formatting the document"),
}).strict().optional()

export const FormatterConfigSchema = z.object({
  enabled: z.boolean().optional().describe("Whether the formatter is enabled"),
  include: z.array(z.string()).optional().describe("Additional glob patterns to include beyond defaults (e.g., ['**/*.xml.erb', 'custom/**/*.html'])"),
  exclude: z.array(z.string()).optional().describe("Glob patterns to exclude from formatting"),
  indentWidth: z.number().int().positive().optional().describe("Number of spaces per indentation level"),
  indentStyle: z.enum(["space", "tab"]).optional().describe("Indentation character to use ('space' or 'tab')"),
  maxLineLength: z.number().int().positive().optional().describe("Maximum line length before wrapping"),
  rewriter: RewriterConfigSchema.describe("Rewriter configuration for pre and post-format transformations"),
}).strict().optional()

export const FrameworkSchema = z.enum(FRAMEWORK_NAMES).optional()

export const EnvironmentSchema = z.enum(ENVIRONMENT_NAMES)
  .describe("Framework context (default: 'ruby')")

export const TemplateEngineSchema = z.enum(["erubi", "erb", "herb"]).optional()
  .describe("Template engine used for compilation (default: 'erubi')")

export const EngineConfigSchema = z.record(z.string(), z.unknown()).nullish()

export const HerbConfigSchema = z.object({
  version: z.string().describe("Configuration file version"),
  framework: FrameworkSchema,
  template_engine: TemplateEngineSchema,
  files: FilesConfigSchema.describe("Top-level file configuration"),
  engine: EngineConfigSchema.describe("Engine configuration"),
  linter: LinterConfigSchema,
  formatter: FormatterConfigSchema,
}).strict()

export type HerbConfigSchemaType = z.infer<typeof HerbConfigSchema>
export type RuleConfigSchemaType = z.infer<typeof RuleConfigSchema>
export type FilesConfigSchemaType = z.infer<typeof FilesConfigSchema>
export type SeveritySchemaType = z.infer<typeof SeveritySchema>

export type SeverityConfig = DiagnosticSeverity | { editor: DiagnosticSeverity; cli: DiagnosticSeverity }

export type LinterMode = "editor" | "cli"

export function resolveSeverity(severity: SeverityConfig, mode: LinterMode): DiagnosticSeverity {
  if (typeof severity === "string") {
    return severity
  }

  return severity[mode]
}

/**
 * Pseudo rule name used inside `linter.rules` to set the default `enabled`
 * state for every rule that isn't explicitly configured.
 *
 * ```yaml
 * linter:
 *   rules:
 *     all:
 *       enabled: false
 *
 *     html-no-event-handlers:
 *       enabled: true
 * ```
 */
export const ALL_RULES_KEY = "all"
